modules/gapi/src/compiler/gcompiler.cpp

```diff
@@ -317,14 +317,24 @@ void cv::gimpl::GCompiler::validateInputMeta()
                      "got " + std::to_string(m_metas.size()) + " meta arguments)"));
     }
 
-    const auto meta_matches = [](const GMetaArg &meta, const GProtoArg &proto) {
+    const auto meta_matches = [](const GMetaArg &meta, const GProtoArg &proto, std::ostream& tracer) {
         switch (proto.index())
         {
         // FIXME: Auto-generate methods like this from traits:
         case GProtoArg::index_of<cv::GMat>():
         case GProtoArg::index_of<cv::GMatP>():
-            return util::holds_alternative<cv::GMatDesc>(meta);
-
+        {
+            bool is_valid = util::holds_alternative<cv::GMatDesc>(meta);
+            if (is_valid)
+            {
+                if(cv::empty_gmat_desc() == cv::util::get<cv::GMatDesc>(meta))
+                {
+                    is_valid = false;
+                    tracer << "empty cv::Mat is not allowed as compile argument";
+                }
+            }
+            return is_valid;
+        }
```

Comment:
We don't follow (say, respect) the rule of single return point so this construct can be simplified like:

```c++
if (!util::holds_alternative<cv::GMatDesc>(meta)) {
    return false;
}
if (cv::empty_gmat_desc() == cv::util::get<cv::GMatDesc>(meta)) {
    complain
    return false;
}
return true; // checks passed
```

Reply:
fixed